semgrep-core/src/analyzing/Constant_propagation.ml

```diff
@@ -686,4 +686,12 @@ let propagate_dataflow lang ast =
                 propagate_dataflow_one_function lang inputs flow);
           }
       in
+      (* Hack to do dataflow on the entire AST.
+         Just think of it as a function body for a function which takes no arguments.
+         
+         Shouldn't repeat work, because the overall dataflow engine doesn't know about the inside of functions. (To be changed?)
+      *)
+      let stmts = List.concat_map (AST_to_IL.stmt lang) ast in
```

Comment:
Isn't this done in a better way with your new proposal?

Here it seems that if we had:

```
x = 1;
function f() {
  x = 2;
}
foo(x);
function g() {
  x = 3;
}
bar(x);
```
this would be interpreted as
```
x = 1;
x = 2;
foo(x); # would incorrectly infer that this is `foo(2)`
x = 3;
bar(x); # would incorrectly infer that this is `bar(3)`
```
am I correct?

Reply:
Sorry no, indeed this should work. I got confused by the `List.concat_map`, but this should be equivalent to `vAST_to_IL.stmt lang (G.stmt1 ast)`. However I still think your new proposal would cover this already, or doesn't it?